receiver/telegrafreceiver/converter.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2021, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package telegrafreceiver
+
+import (
+	"fmt"
+
+	"github.com/influxdata/telegraf"
+	"go.opentelemetry.io/collector/consumer/pdata"
+)
+
+const (
+	fieldLabel = "field"
+)
+
+type MetricConverter interface {
+	Convert(telegraf.Metric) (pdata.Metrics, error)
+}
+
+type metricConverter struct {
+	separateField bool
+}
+
+func newConverter(separateField bool) MetricConverter {
+	return metricConverter{
+		separateField: separateField,
+	}
+}
+
+// Convert converts telegraf.Metric to pdata.Metrics.
+func (mc metricConverter) Convert(m telegraf.Metric) (pdata.Metrics, error) {
+	ms := pdata.NewMetrics()
+	rms := ms.ResourceMetrics()
+	rms.Resize(1)
+	rm := rms.At(0)
+
+	// Attach tags as attributes - pipe through the metadata
+	for _, t := range m.TagList() {
+		rm.Resource().Attributes().InsertString(t.Key, t.Value)
+	}
+
+	rm.InstrumentationLibraryMetrics().Resize(1)
+	ilm := rm.InstrumentationLibraryMetrics().At(0)
+
+	il := ilm.InstrumentationLibrary()
+	il.SetName(typeStr)
+	il.SetVersion(versionStr)
+
+	tim := m.Time().UnixNano()
+
+	metrics := ilm.Metrics()
+
+	switch t := m.Type(); t {
```

Comment:
Just thinking about this in the context of our Zoom conversation earlier.

The Prometheus Telegraf input plugin is the only Telegraf input plugin that sets this type field. Everything else defaults to "untyped". To be sure, that Telegraf plugin is very popular, but otelcol already has a Prometheus receiver (or two), so those users won't benefit much from this switch block.

I talked with someone on the Telegraf team, and we agreed that Gauge is probably the best we can do here.